### ISSUE TITLE
create a new unique identifier per book with hash of the epub file

### ISSUE THREAD

**Issue by panaC** [readium/readium-desktop#789]
https://stackoverflow.com/questions/10070293/choosing-a-hash-function-for-best-performance#10074975

https://automationrhapsody.com/md5-sha-1-sha-256-sha-512-speed-performance/

https://github.com/hex7c0/nodejs-hash-performance

https://nodejs.org/api/crypto.html#crypto_determining_if_crypto_support_is_unavailable

**Comment by danielweck** [comment]
I don't think we need to hash full EPUB files (even with fast CRC32-like techniques), because publications vary in size enormously (+10MB, +100MB sometimes)

```
import * as crypto from "crypto";

const uniqueStr = ; // for example, concatenate the contents of `book.zip!/EPUB/package.opf` and the EPUB file size (or some other unique-ish metadata, but not filename / filesystem location / originating URL, etc.)

const checkSum = crypto.createHash("sha256");
checkSum.update(uniqueStr);
const sha256Hex = checkSum.digest("hex");

// then use `sha256Hex` as a fixed-length unique ID for the EPUB
```

**Comment by danielweck** [comment]
General discussions: https://github.com/readium/architecture/issues/68

**Comment by danielweck** [comment]
Isn't this issue a duplicate of https://github.com/readium/readium-desktop/issues/700
?

**Comment by danielweck** [comment]
Let's close this issue and continue discussing in the older one:
https://github.com/readium/readium-desktop/issues/700#issuecomment-534023267
(I have already copied across the comments from here)
